compiler/layouter.nim

```diff
@@ -145,6 +145,10 @@ proc closeEmitter*(em: var Emitter) =
       maxLhs = 0
       lineBegin = i+1
     of ltTab:
+      if i >= 2 and em.kinds[i-2] == ltNewline and em.kinds[i-1] in {ltNewline, ltSpaces}:
```

Comment:
I don't understand what this patch has to do with the added test cases.

Reply:
This is the first test case:
```nim
block:
  var t = 3

## This MUST be a multiline comment,
## single line comment would be ok.
block:
  var x = 7
```

Tokens for the important part of the test case look like this:
```
...
(token: "=", kind: ltOther, maxLhs: 1)
(token: " ", kind: ltSpaces, maxLhs: 1)
(token: "3", kind: ltLit, maxLhs: 1)
(token: "\n", kind: ltNewline, maxLhs: 1)    <--------------------
(token: "\n", kind: ltNewline, maxLhs: 1)    <--------------------
(token: "", kind: ltTab, maxLhs: 1)
(token: "## This MUST be a multiline comment,", kind: ltComment, maxLhs: 1)
(token: "\n", kind: ltNewline, maxLhs: 1)
(token: "", kind: ltTab, maxLhs: 1)
(token: "## single line comment would be ok.", kind: ltComment, maxLhs: 1)
(token: "", kind: ltEndSection, maxLhs: 1)
(token: "\n", kind: ltNewline, maxLhs: 0)
(token: "block", kind: ltKeyword, maxLhs: 0)
(token: ":", kind: ltOther, maxLhs: 0)
...
```

As you can see, `ltEndSection` is reached too late (after the doc comment) so the doc comment still has `maxLhs = 1`, which means that nimpretty produces the following:

```nim
block:
  var t = 3

 ## This MUST be a multiline comment,
 ## single line comment would be ok.
block:
  var x = 7
```

Notice the one-space indentation forthe doc comment, which produces `Error: invalid indentation` and the file doesn't compile anymore.

----

The other test case:
```nim
block:
  var t = 3
  ## another
  ## multi

## This MUST be a multiline comment,
## single line comment would be ok.
block:
  var x = 7
```

has the following structure:
```
...
(token: "=", kind: ltOther, maxLhs: 3)
(token: " ", kind: ltSpaces, maxLhs: 3)
(token: "3", kind: ltLit, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 3)    <--------------------
(token: "  ", kind: ltSpaces, maxLhs: 3)     <--------------------
(token: "", kind: ltTab, maxLhs: 3)
(token: "## another", kind: ltComment, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 3)
(token: "  ", kind: ltTab, maxLhs: 3)
(token: "## multi", kind: ltComment, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 3)
(token: "", kind: ltTab, maxLhs: 3)
(token: "## This MUST be a multiline comment,", kind: ltComment, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 3)
(token: "", kind: ltTab, maxLhs: 3)
(token: "## single line comment would be ok.", kind: ltComment, maxLhs: 3)
(token: "", kind: ltEndSection, maxLhs: 3)
(token: "\n", kind: ltNewline, maxLhs: 0)
(token: "block", kind: ltKeyword, maxLhs: 0)
(token: ":", kind: ltOther, maxLhs: 0)
...
```

Nimpretty produces:
```nim
block:
  var t = 3
   ## another
   ## multi

   ## This MUST be a multiline comment,
   ## single line comment would be ok.
block:
  var x = 7
```

Notice that all comments have 3-space indentation, which also breaks the compilation.

----

With this patch, we check if there is an empty line before `ltTab` (the first test case) or the contents are already aligned (the second test case), and in those cases we reset `maxLhs` and consequently don't add extra spaces.